File: CS50x/Problem Set 6/sentimental-credit/credit.py
import sys

def main():
    card = get_card()
    length = len(card)
    if length not in [13, 15, 16]:
        print("INVALID")
        sys.exit(1)
    #luhn(card_to_int(card))
    if luhn(card) == False:
        print("INVALID")
        sys.exit(2)
    if length in [13, 16] and card[0] == "4":
        print("VISA")
        sys.exit(0)
    if length == 15 and card[0]+card[1] in ["34", "37"]:
        print("AMEX")
        sys.exit(0)
    if length == 16 and card[0] == "5" and 1 <= int(card[1]) <= 5:
        print("MASTERCARD")
        sys.exit(0)
    print("INVALID")
    sys.exit(3)

def luhn(card):
    i = -2
    x = 0
    y = 0
    while i >= -len(card):
        xx = 2 * int(card[i])
        for val in str(xx):
            x += int(val)
        i -= 2
    i = -1
    while i >= -len(card):
        y += int(card[i])
        i -= 2
    luhnval = x + y
    if luhnval % 10 == 0:
        return True
    return False


#def card_to_int(card):
#    i = 0
#    int_card = 0
#    for number in reversed(card):
#        int_card += int(number)*pow(10, i)
#        i += 1
#    return int_card


def get_card():
    while 1:
        card = input("Number: ").strip()
        error = False
        for num in card:
            try:
                int(num)
            except:
                error = True
        if (error == False):
            return card

if __name__ == "__main__":
    main()

File: CS50x/Problem Set 1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int digits(long num, int place);
int length(long num);
void raise(int type);
int luhn(long num);

int main(void){
    long card = get_long("Card: ");
    int l = length(card);
    if(l < 13 || l > 16){
        raise(0);
    }
    else if(l == 13 && digits(card, 13) == 4){
        raise(luhn(card)*2);
    }
    else if(l == 16){ // dont need length, if digits too high will be 0
        if(digits(card, 16) == 4){
            raise(luhn(card)*2);
        }
        else if(digits(card, 16) == 5){
            if(digits(card, 15) > 0 && digits(card, 15) < 6){
                raise(luhn(card)*3);
            }
            else{
                raise(0);
            }
        }
        else{
            raise(0);
        }
    }
    else if(l == 15 && digits(card, 15) == 3){ // dont need length == 15, if digits too high will be 0
        if(digits(card, 14) == 4 || digits(card, 14) == 7){
            raise(luhn(card)*4);
        }
        else{
            raise(0);
        }
    }
    else{
        raise(0);
    }
}


int digits(long num, int place){
    return num % (long)pow(10, place) / (long)pow(10,place-1);
}


int length(long num){
    int x = 0;
    while(num > 0) {
        num = num / 10;
        x ++;
    }
    return x;
}

void raise(int type){
    if(type == 0){
        printf("INVALID\n");
    }
    else if (type == 1){
        printf("INVALID\n");
    }
    else if (type == 2){
        printf("VISA\n");
    }
    else if (type == 3){
        printf("MASTERCARD\n");
    }
    else if (type == 4){
        printf("AMEX\n");
    }
}

int luhn(long num){
    int x = 0;
    for(int i = 2; i < 17; i = i + 2){
        int xx = 2 * digits(num, i);
        x = x + digits((long)xx, 1) + digits((long)xx, 2);
        printf("%i\n", x);
    }
    for(int j = 1; j < 16; j = j + 2){
        int y = digits(num, j);
        x = x + y;
        printf("%i\n", x);
    }
    if(digits(x, 1) == 0){
        return 1;
    }
    else{
        return 0;
    }
}
